common/include/window.h

```diff
@@ -50,6 +50,79 @@ static inline void set_embedded_window_pointer(embed_window_pointer_t *wp, windo
 
 static inline void set_embedded_window_pointer(ignore_window_pointer_t *, window *) {}
 
+struct window
+{
+private:
+	grs_canvas w_canv;					// the window's canvas to draw to
+	window_event_result (*w_callback)(window *wind,const d_event &event, void *data);	// the event handler
+	int w_visible;						// whether it's visible
+	int w_modal;						// modal = accept all user input exclusively
+	void *data;							// whatever the user wants (eg menu data for 'newmenu' menus)
+	struct window *prev;				// the previous window in the doubly linked list
+	struct window *next;				// the next window in the doubly linked list
+	
+public:
+	// Declaring as friends to keep function syntax, for historical reasons (for now at least)
+	friend window *window_create(grs_canvas *src, int x, int y, int w, int h, window_subfunction<void> event_callback, void *userdata, const void *createdata);
+	
+	friend int window_close(window *wind);
+	friend int window_exists(window *wind);
+	friend window *window_get_front();
+	friend window *window_get_first();
+	friend void window_select(window &wind);
+	friend window *window_set_visible(window &wind, int visible);
+#if !DXX_USE_OGL
+	friend void window_update_canvases();
+#endif
+	friend window_event_result window_send_event(window &wind,const d_event &event);
+	friend void window_set_modal(window &wind, int modal);
+	friend int window_is_modal(window &wind);
+	
+	friend grs_canvas &window_get_canvas(window &wind)
+	{
+		return wind.w_canv;
+	}
+	
+	friend window *window_set_visible(window *wind, int visible)
+	{
+		return window_set_visible(*wind, visible);
+	}
+
+	friend int window_is_visible(window *wind)
```

Comment:
Yes, it is enforced only at compile time.  In some cases, a static analysis tool might warn if it discovers a code path where `nullptr` will be dereferenced and passed as a reference, but that is not required of a conforming compiler, nor is it possible in all cases.

I agree that this should not attempt to update all the pointer form calls at once.  Rather, I want to allow the existing reference form calls to retain that form, with a goal of eventually converting other places to use the reference form.  For now, having both forms and allowing the pointer form not to be checked at compile time is fine.